File: morphac/simulation/test/playground_state_test.cc
#include "Eigen/Dense"
#include "gtest/gtest.h"

#include "environment/include/map.h"
#include "mechanics/models/include/diffdrive_model.h"
#include "simulation/include/playground_state.h"

namespace {

using std::make_unique;
using std::srand;
using std::unique_ptr;

using Eigen::MatrixXd;

using morphac::environment::Map;
using morphac::mechanics::models::DiffDriveModel;
using morphac::mechanics::models::KinematicModel;
using morphac::robot::blueprint::Footprint;
using morphac::robot::blueprint::Robot;
using morphac::simulation::PlaygroundState;

class PlaygroundStateTest : public ::testing::Test {
 protected:
  PlaygroundStateTest() {
    // Set random seed for Eigen.
    srand(7);

    playground_state1_ = make_unique<PlaygroundState>(Map(40, 20, 0.1));
    playground_state2_ = make_unique<PlaygroundState>(Map(map_data_, 0.1));
    DiffDriveModel diffdrive_model1(1., 1.);
    DiffDriveModel diffdrive_model2(2., 2.);
    robot1_ =
        make_unique<Robot>(diffdrive_model1, Footprint(MatrixXd::Zero(10, 2)));
    robot2_ =
        make_unique<Robot>(diffdrive_model2, Footprint(MatrixXd::Ones(10, 2)));
  }

  void SetUp() override {}

  unique_ptr<PlaygroundState> playground_state1_, playground_state2_;
  unique_ptr<Robot> robot1_, robot2_;
  MatrixXd map_data_ = MatrixXd::Random(300, 300);
};

TEST_F(PlaygroundStateTest, GetMap) {
  ASSERT_TRUE(playground_state1_->get_map().get_data().isApprox(
      MatrixXd::Zero(200, 400)));
  ASSERT_EQ(playground_state1_->get_map().get_width(), 40.);
  ASSERT_EQ(playground_state1_->get_map().get_height(), 20.);
  ASSERT_EQ(playground_state1_->get_map().get_resolution(), 0.1);

  ASSERT_TRUE(playground_state2_->get_map().get_data().isApprox(map_data_));
  ASSERT_EQ(playground_state2_->get_map().get_width(), 30.);
  ASSERT_EQ(playground_state2_->get_map().get_height(), 30.);
  ASSERT_EQ(playground_state2_->get_map().get_resolution(), 0.1);
}

TEST_F(PlaygroundStateTest, SetMap) {
  MatrixXd map_data = MatrixXd::Random(400, 200);
  playground_state1_->set_map(Map(map_data, 0.1));
  ASSERT_TRUE(playground_state1_->get_map().get_data().isApprox(map_data));

  playground_state2_->set_map(Map(30., 30., 0.1));
  ASSERT_TRUE(playground_state2_->get_map().get_data().isApprox(
      MatrixXd::Zero(300, 300)));
}

TEST_F(PlaygroundStateTest, AddRobot) {
  playground_state1_->AddRobot(*robot1_, 0);
  ASSERT_EQ(playground_state1_->NumRobots(), 1);

  playground_state2_->AddRobot(*robot1_, 0);
  playground_state2_->AddRobot(*robot2_, 1);
  ASSERT_EQ(playground_state2_->NumRobots(), 2);
}

TEST_F(PlaygroundStateTest, InvalidAddRobot) {
  // Add robot with invalid UIDs
  ASSERT_THROW(playground_state1_->AddRobot(*robot1_, -1),
               std::invalid_argument);

  // Add robot with duplicate UIDs
  playground_state2_->AddRobot(*robot1_, 0);
  ASSERT_THROW(playground_state2_->AddRobot(*robot2_, 0),
               std::invalid_argument);

  // It should be possible to add the same robot object multiple times.
  playground_state2_->AddRobot(*robot1_, 1);
}

TEST_F(PlaygroundStateTest, GetRobotOracle) {
  // Adding robots.
  playground_state1_->AddRobot(*robot1_, 0);
  playground_state1_->AddRobot(*robot2_, 1);

  auto robot_oracle = playground_state1_->get_robot_oracle();

  // Making sure that the oracle has the right size
  ASSERT_EQ(robot_oracle.size(), 2);

  // Testing actual contents of the oracle. It must store the right robots
  // against the right uids
  //const DiffDriveModel &k =
  //    dynamic_cast<const DiffDriveModel &>(robot1_->get_kinematic_model());

  const KinematicModel& k = robot1_->get_kinematic_model();
  std::cout << k.pose_size << std::endl;

  ASSERT_TRUE(robot_oracle.find(0)->second.get_footprint().get_data().isApprox(
      MatrixXd::Zero(10, 2)));
  // ASSERT_EQ(dynamic_cast<DiffDriveModel &>(
  //              const_cast<KinematicModel &>(
  //                  robot_oracle.find(0)->second.get_kinematic_model()))
  //              .radius,
  //          1.);
}

}  // namespace

int main(int argc, char **argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: morphac/simulation/python/test/playground_state_test.py
import numpy as np
import pytest

from morphac.constructs import State
from morphac.environment import Map
from morphac.mechanics.models import DiffDriveModel
from morphac.robot.blueprint import Footprint, Robot
from morphac.simulation import PlaygroundState


@pytest.fixture()
def generate_playground_state_list():
    ps1 = PlaygroundState(Map(10., 10., 0.1))
    ps2 = PlaygroundState(map=Map(np.ones([200, 200], dtype=np.float), 0.1))

    return ps1, ps2


@pytest.fixture()
def generate_robot_list():
    r1 = Robot(DiffDriveModel(1., 1.), Footprint([[0, 0]]))
    r2 = Robot(DiffDriveModel(2., 3.), Footprint(
        [[0, 0]]), State([1., 2., 3.], []))

    return r1, r2


def test_map(generate_playground_state_list):
    ps1, ps2 = generate_playground_state_list

    # Test map retrieval.
    assert ps1.map.width == 10.
    assert ps1.map.height == 10.
    assert ps1.map.resolution == 0.1
    assert np.all(ps1.map.data == 0.)

    assert ps2.map.width == 20.
    assert ps2.map.height == 20.
    assert ps2.map.resolution == 0.1
    assert np.all(ps2.map.data == 1.)

    # Test map setting.
    ps1.map = Map(10., 10., 0.5)
    ps2.map = Map(20., 20., 0.8)

    assert ps1.map.resolution == 0.5
    assert ps2.map.resolution == 0.8
    assert np.all(ps2.map.data == 0.)


def test_add_robot(generate_playground_state_list, generate_robot_list):
    ps1, _ = generate_playground_state_list
    r1, r2 = generate_robot_list

    # Adding the robots.
    ps1.add_robot(r1, 0)

    # Make sure that the robot has been added.
    assert ps1.num_robots == 1

    # Adding another robot and testing.
    ps1.add_robot(r2, 1)

    assert ps1.num_robots == 2

    # Test addition of robots with invalid or duplicate uids.
    with pytest.raises(ValueError):
        ps1.add_robot(r1, -1)
    with pytest.raises(ValueError):
        ps1.add_robot(r2, 0)
    with pytest.raises(ValueError):
        ps1.add_robot(r2, 1)


def test_robot_oracle(generate_playground_state_list, generate_robot_list):
    ps1, _ = generate_playground_state_list
    r1, r2 = generate_robot_list

    # Adding both robots.
    ps1.add_robot(r1, 1)
    ps1.add_robot(r2, 2)

    robot_oracle = ps1.robot_oracle

    # Making sure that its a dict and has the UIDs as keys.
    assert isinstance(robot_oracle, dict)
    assert 1 in robot_oracle
    assert 2 in robot_oracle

    # Checking the size.
    assert len(robot_oracle) == 2

    # Test that the correct robots are stored agains the UIDs.
    assert np.allclose(robot_oracle[1].state.data, [0, 0, 0])
    assert np.allclose(robot_oracle[2].state.data, [1, 2, 3])


def test_get_robot(generate_playground_state_list, generate_robot_list):
    ps1, _ = generate_playground_state_list
    r1, r2 = generate_robot_list

    # Adding both robots.
    ps1.add_robot(r1, 1)
    ps1.add_robot(r2, 2)

    # Test that the correct robots can be retrieved.
    assert np.allclose(ps1.get_robot(1).state.data, [0, 0, 0])
    assert np.allclose(ps1.get_robot(2).state.data, [1, 2, 3])

    # Trying to get a robot against an invalid uid must throw an exception.
    with pytest.raises(ValueError):
        ps1.get_robot(-1)
    with pytest.raises(ValueError):
        ps1.get_robot(3)


def test_get_robot_state(generate_playground_state_list, generate_robot_list):
    ps1, _ = generate_playground_state_list
    r1, r2 = generate_robot_list

    # Adding both robots.
    ps1.add_robot(r1, 1)
    ps1.add_robot(r2, 2)

    # Test that the correct robot states can be retrieved.
    assert np.allclose(ps1.get_robot_state(1).data, [0, 0, 0])
    assert np.allclose(ps1.get_robot_state(2).data, [1, 2, 3])

    # Trying to get a robot state against an invalid uid must throw an
    # exeption.
    with pytest.raises(ValueError):
        ps1.get_robot_state(-1)
    with pytest.raises(ValueError):
        ps1.get_robot_state(3)


def test_set_robot_state(generate_playground_state_list, generate_robot_list):
    ps1, _ = generate_playground_state_list
    r1, r2 = generate_robot_list

    # Adding both robots.
    ps1.add_robot(r1, 1)
    ps1.add_robot(r2, 2)

    # Setting different states.
    ps1.set_robot_state(State([-1, -2, -3], []), 1)
    ps1.set_robot_state(State([1, 1, 2], []), 2)

    # Make sure that the states have updated, through the oracle, robot and
    # robot states.
    robot_oracle = ps1.robot_oracle
    assert np.allclose(robot_oracle[1].state.data, [-1, -2, -3])
    assert np.allclose(robot_oracle[2].state.data, [1, 1, 2])
    assert np.allclose(ps1.get_robot(1).state.data, [-1, -2, -3])
    assert np.allclose(ps1.get_robot(2).state.data, [1, 1, 2])
    assert np.allclose(ps1.get_robot_state(1).data, [-1, -2, -3])
    assert np.allclose(ps1.get_robot_state(2).data, [1, 1, 2])

    # Trying to set a robot state against an invalid uid must throw an
    # exception.
    with pytest.raises(ValueError):
        ps1.set_robot_state(State(3, 0), -1)
    with pytest.raises(ValueError):
        ps1.set_robot_state(State(3, 0), 3)
